jax_triton/pallas/pallas_call.py

```diff
@@ -221,42 +269,110 @@ def clear_caches():
+  if grid is None:
+    return (1,)
+  if isinstance(grid, int):
+    return (grid,)
+  return grid
+
+def _extract_function_name(f: Callable, name: Optional[str]) -> str:
+  if name is None:
+    name = f.__name__ if hasattr(f, "__name__") and f.__name__ else "func"
+  return name
+
+def _convert_block_spec_to_block_mapping(
+    grid: Grid, block_spec: Optional[BlockSpec]) -> Optional[BlockMapping]:
+  if block_spec is None:
+    return None
+  in_avals = [jax_core.ShapedArray((), jnp.int32) for _ in grid]
+  jaxpr, _, consts = pe.trace_to_jaxpr_dynamic(
+      lu.wrap_init(block_spec.index_map), in_avals)
+  if consts:
+    raise NotImplementedError
+  return BlockMapping(block_spec.block_shape, jax_core.ClosedJaxpr(jaxpr, consts))
+
+def pallas_call(f: Callable, out_shape: Any, *, debug: bool = False,
+                grid: Optional[Grid] = None,
+                in_specs: Optional[Sequence[Optional[BlockSpec]]] = None,
+                out_specs: Optional[Sequence[Optional[BlockSpec]]] = None,
                 num_warps: int = 4, num_stages: int = 3,
                 input_output_aliases: Dict[int, int] = {},
                 interpret: bool = False,
-                name: Optional[str] = None,
-                **metaparams):
-  if isinstance(grid, int):
-    grid = (grid,)
-  if name is not None:
-    name = name
-  else:
-    name = f.__name__ if hasattr(f, "__name__") and f.__name__ else "func"
+                name: Optional[str] = None):
+  grid = _preprocess_grid(grid)
+  name = _extract_function_name(f, name)
+  singleton = False
+  if not isinstance(out_shape, (tuple, list)):
+    out_shape = (out_shape,)
+    singleton = True
   flat_out_shapes, out_tree = tree_util.tree_flatten(out_shape)
-  flat_out_shapes = [jax.ShapeDtypeStruct(x.shape, x.dtype) for x in
-                     flat_out_shapes]
-  def wrapped(*args, **kwargs):
-    flat_args, _ = tree_util.tree_flatten((args, kwargs))
-    in_tree = tree_util.tree_structure((*args, *flat_out_shapes))
-    flat_avals = map(
-        jax_core.raise_to_shaped, map(jax_core.get_aval, flat_args))
-    flat_output_avals = [jax_core.ShapedArray(out_shape.shape, out_shape.dtype)
-        for out_shape in flat_out_shapes]
-    ptr_avals = [state.ShapedArrayRef(aval.shape, aval.dtype) for aval in flat_avals]
-    out_ptr_avals = [state.ShapedArrayRef(aval.shape, aval.dtype) for aval in flat_output_avals]
+  if out_specs is not None:
+    out_specs = tree_util.tree_unflatten(
+        out_tree, tree_util.tree_leaves(out_specs))
+  flat_out_shapes = [jax.ShapeDtypeStruct(x.shape, x.dtype)
+                     for x in flat_out_shapes]
+  @jax.jit
+  def wrapped(*args):
+    flat_args, in_tree = tree_util.tree_flatten(args)
+    if grid is None:
+      if in_specs is not None:
+        raise ValueError("Cannot specify `in_specs` with a `None` grid.")
+      if out_specs is not None:
+        raise ValueError("Cannot specify `out_specs` with a `None` grid.")
```

Comment:
You can perform those two checks outside of `wrapped` to validate the arguments when `pallas_call` is used to wrap a function, not when it's called the first time.

Reply:
Done.